### PR TITLE
Add an assume that the index is inbounds to slice::get_unchecked

### LINKED ISSUE #116878: Missed optimization with `slice::get_unchecked` followed by `<slice as Index>::index` at a lower index
I tried this code:

```rust
pub unsafe fn cannot_elide_bounds_check(s: &[u8]) -> u8 {
    let a = *s.get_unchecked(1);
    a + s[0]
}
pub unsafe fn can_elide_bounds_check(s: &[u8]) -> u8 {
    let a = *s.get(1).unwrap_unchecked();
    a + s[0]
}
```

Compile with `-Copt-level=3` on godbolt.org ([godbolt link](https://rust.godbolt.org/z/qvzEEzEn7))

I expected to see this happen: Both functions should compile to approximately the same thing, with no bounds check on the `s[0]` since the optimizer should be able to see that the slice is at least of length 2 due to the previous unchecked indexing.

```asm
example::can_elide_bounds_check:
        movzx   eax, byte ptr [rdi]
        add     al, byte ptr [rdi + 1]
        ret
```

Instead, this happened: `cannot_elide_bounds_check` contains a bounds check (only `can_elide_bounds_check` does not contain a bounds check).

```asm
example::cannot_elide_bounds_check:
        test    rsi, rsi
        je      .LBB0_2
        movzx   eax, byte ptr [rdi]
        add     al, byte ptr [rdi + 1]
        ret
.LBB0_2:
        push    rax
        lea     rdx, [rip + .L__unnamed_1]
        xor     edi, edi
        xor     esi, esi
        call    qword ptr [rip + core::panicking::panic_bounds_check@GOTPCREL]
        ud2
```

### Meta
rustc stable (1.73.0), nightly (2023-10-17), and 1.58.0 (when `Option::unwrap_unchecked` was stabilized) all have approximately the same codegen.
`rustc --version --verbose`:
```
rustc 1.75.0-nightly (09df6108c 2023-10-17)
binary: rustc
commit-hash: 09df6108c84fdec400043d99d9ee232336fd5a9f
commit-date: 2023-10-17
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.2
```

(no backtrace)

@rustbot label +C-optimization -C-bug

### PR DESCRIPTION
**Pull request by bend-n**
Fixes #116878

### CODE REVIEW

File: library/core/src/slice/mod.rs [hunk 1]
```diff
@@ -662,7 +662,7 @@ impl<T> [T] {
         // SAFETY: the caller must uphold most of the safety requirements for `get_unchecked`;
         // the slice is dereferenceable because `self` is a safe reference.
         // The returned pointer is safe because impls of `SliceIndex` have to guarantee that it is.
-        unsafe { &*index.get_unchecked(self) }
+        unsafe { &*index.get(self).unwrap_unchecked() }
```
Comment (oli-obk) [inline library/core/src/slice/mod.rs:665]: Would we get the same result by changing SliceIndex::get_unchecked instead of just the slice method?

Comment (bend-n) [reply library/core/src/slice/mod.rs:665]: Probably? Will check.

File: src/tools/miri/tests/fail/stacked_borrows/zst_slice.rs [hunk 2]
```diff
@@ -1,5 +1,5 @@
 //@compile-flags: -Zmiri-strict-provenance
-//@error-in-other-file: /retag .* tag does not exist in the borrow stack/
+//@error-in-other-file: unreachable
```
Comment (RalfJung) [inline src/tools/miri/tests/fail/stacked_borrows/zst_slice.rs:2]: You just made this test useless. It's a Stacked Borrows test, it must lead to a Stacked Borrows error. So this definitely cannot land.

File: library/core/src/slice/mod.rs (hunk 1 above)
Comment (oli-obk) [reply library/core/src/slice/mod.rs:665]: We can change that, so that `get` does the full logic and `get_unchecked` does the change from this PR

**Comment by oli-obk** [comment]
looks like the large regression is in LLVM, probably due to more optimizations triggering?

**Comment by Nilstrieb** [comment]
> probably due to more optimizations triggering?

could also just be the additional work of the extra abstractions

**Comment by saethlin** [comment]
The increased complexity of the optimized MIR backs up that conclusion: https://godbolt.org/z/Kned9hxx6

I strongly suspect an implementation with `assume` would look better in perf and produce the same optimizations, because LLVM eventually manages to optimzie the `.get().unwrap_unchecked()` into an `assume`.

**Comment by bend-n** [comment]
Im trying it with the `unwrap_unchecked` in `SliceIndex<usize>::get_unchecked`, then we can try a assume? Just to have more to compare with?

**Comment by saethlin** [comment]
There's no harm in trying all the ideas, especially when the perf queue isn't very busy :)

File: library/core/src/slice/index.rs [hunk 3]
```diff
@@ -214,7 +214,8 @@ unsafe impl<T> SliceIndex<[T]> for usize {
     #[inline]
     fn get(self, slice: &[T]) -> Option<&T> {
         // SAFETY: `self` is checked to be in bounds.
-        if self < slice.len() { unsafe { Some(&*self.get_unchecked(slice)) } } else { None }
+        // The call to `add` is safe, as we check the bounds.
+        if self < slice.len() { unsafe { Some(&*slice.as_ptr().add(self)) } } else { None }
```
Comment (Nilstrieb) [inline library/core/src/slice/index.rs:218]: this could be a perf win on its own, so testing this together with the other change isn't super useful.

Comment (bend-n) [reply library/core/src/slice/index.rs:218]: Do you want to try that change on its own?

**Comment by bend-n** [comment]
If im reading the [results](https://perf.rust-lang.org/compare.html?start=3fbcfd2b6f7030cb70328aa759107efc1516912c&end=f5944a25b80d8f8598e7719818bf18740b2adada&stat=instructions%3Au&tab=compile) correctly, it seems 2073441e02019e32078091303eb4cf2af23bf04e (unwrap_unchecked) is... better than 2bb9423e048c3d4296ed8cd1c80eca22ec7820be (assume).

Is that because of the difference in `SliceIndex<usize>::get`?
Why is the test failing, but I cant make it fail locally?

Also, is instruction count the number of instructions in the binary? Is there a way to see the diff of the instructions? Like, are there more instructions because its getting better codegen, or is it just throwing a wrench at the optimizer?

**Comment by saethlin** [comment]
I think you are reading the results wrong. All the try builds on this PR have a different parent commit, which is whatever the tip of master was then the build was done. You need to compare each of them to their own try build, which the links in the perf bot comments do for you.

All the perf runs in this PR look the same to me within noise. If I had to guess, the primary effect here is LLVM doing more work in optimizing some code path that `regex` uses a lot. Unclear if this leads to better codegen, but the regression in compile time here is marginal and I think you demonstrated with this codegen test that we do optimize better with this PR.

The instruction count is the number of executed instructions. There is no diff view available. You can locally generate a cachegrind instruction diff but it is only instructions per function, and inlining is not accounted for so the view is *very* coarse. If you click on any of the rows in the perf report it'll show you the command to generate a cachegrind diff in a checkout of rust-lang/rustc-perf.

If you want to understand the codegen differences I would see if `regex` has benchmarks and poke around those. See if they get faster with this PR, and you can do some objdump or whatnot on the bench binaries to see how if at all the codegen changed.

**Comment by saethlin** [comment]
For the rest of the Miri team here: This test is definitely not supposed to fail right? Do you think this indicates a miscompilation?

**Comment by bend-n** [comment]
@rustbot review

isnt this supposed to be superseded by the MIR thing though?

**Comment by saethlin** [comment]
Scott said no over here: https://github.com/rust-lang/rust/pull/117035#issuecomment-1777608934 which I think is the PR you are referring to

Since this fixes a codegen issue, we should have a codegen test for this. The existing ones are under `tests/codegen/`. You'll probably need a combination of `CHECK` and `CHECK-NOT` to specify the sequence of LLVM instructions you're expecting.

**Comment by bend-n** [comment]
we have a codegen test for this

**Comment by saethlin** [comment]
:facepalm: Of course we do.

I'll look over this again in the morning when I haven't been doing Rust for so many hours in a row.

**Comment by saethlin** [comment]
https://github.com/rust-lang/rust/issues/116759
@bors retry

**Comment by nnethercote** [comment]
@saethlin: are the binary size regression expected? They seem surprising to me, given that this change was intended to produce more streamlined code :/

**Comment by bend-n** [comment]
Presumably more optimized code.

**Comment by saethlin** [comment]
The fact that `ripgrep` leads the linked_artifact regressions and that the largest icount regressions are in `regex` makes me think that whatever happened, `regex` is a good example.

The top of the cachegrind diff for regex opt-full is:
```
196,187,152  ???:llvm::ScalarEvolution::isLoopBackedgeGuardedByCond(llvm::Loop const*, llvm::CmpInst::Predicate, llvm::SCEV const*, llvm::SCEV const*) 
177,396,685  ???:llvm::ScalarEvolution::isBasicBlockEntryGuardedByCond(llvm::BasicBlock const*, llvm::CmpInst::Predicate, llvm::SCEV const*, llvm::SCEV const*) 
144,664,351  ???:llvm::computeKnownBitsFromAssume(llvm::Value const*, llvm::KnownBits&, unsigned int, llvm::SimplifyQuery const&)
101,443,164  ???:llvm::ScalarEvolution::isImpliedCond(llvm::CmpInst::Predicate, llvm::SCEV const*, llvm::SCEV const*, llvm::Value const*, bool, llvm::Instruction const*) 
```
Which at least suggests that some of the increased icount in LLVM is processing the assumes. Whether it's helping is less clear.

Just doing an `rg panic_bounds_check target/release/deps/ --count` on builds of `regex-1.10.2` (latest release) before and after this PR indicates that we went from:
```
target/release/deps/memchr-d2bb6823209e6c4b.ll:66
target/release/deps/aho_corasick-dee4e7528de2347a.ll:888
target/release/deps/regex_syntax-cbe8289811e12031.ll:98
target/release/deps/regex_automata-16ffd8f27578710f.ll:1420
```
To
```
target/release/deps/memchr-d2bb6823209e6c4b.ll:66
target/release/deps/aho_corasick-dee4e7528de2347a.ll:888
target/release/deps/regex_syntax-cbe8289811e12031.ll:98
target/release/deps/regex_automata-16ffd8f27578710f.ll:1404
```
Hardly a _massive_ win, but the change does suggest we optimized out some panics that used to be in there.

But grepping for `call` or `invoke` suggests that there were a lot of changes. Those numbers are much larger and change in both directions.

So I have two guesses:
1. LLVM has done some optimization using the assume and that induced further inlining.
2. MIR inlining doesn't ignore assumes in its cost estimation. I don't think it can fully ignore them, because the cost of adding an `assume` is not just the assume itself, but whatever statements are required to produce its predicate. Maybe with SSA you easily back out what statements were used to compute the predicate and ignore those but MIR does not make this easy so I'm going to try just ignoring assumes in MIR inlining and see what happens to perf: https://github.com/rust-lang/rust/pull/118626. If we get the inverse of the perf reports here that would be interesting, but my most realistic expectation is that we get nothing.

**Comment by rylev** [comment]
@bend-n @saethlin do we have proof that this optimization actually makes a positive impact on compiled code? The impact on regex is large enough that it makes me wonder if we should revert and figure out a path forward before moving forward with this. Thoughts?

**Comment by saethlin** [comment]
We have a codegen test in this PR that verifies that it helps

File: library/core/src/slice/index.rs [hunk 4]
```diff
@@ -233,7 +233,10 @@ unsafe impl<T> SliceIndex<[T]> for usize {
         // cannot be longer than `isize::MAX`. They also guarantee that
         // `self` is in bounds of `slice` so `self` cannot overflow an `isize`,
         // so the call to `add` is safe.
-        unsafe { slice.as_ptr().add(self) }
+        unsafe {
+            crate::intrinsics::assume(self < slice.len());
```
[the remark this answers is not in the document]
Comment (saethlin) [reply library/core/src/slice/index.rs:237]: I couldn't figure out how to justify the compile-time regression: https://github.com/rust-lang/rust/pull/120762

[the remark this answers is not in the document]
Comment (saethlin) [reply library/core/src/slice/index.rs:237]: Actually it seems there's a pretty robust effect on the size of `libstd.so` that I just didn't notice before. Hunh.

**Pull request merged**